### PR TITLE
ngcc: configure and update timeouts

### PR DESCRIPTION
**Pull request by petebacondarwin**
Discussions on https://github.com/angular/angular/issues/36796 highlighted that we ought to make the ngcc async locker timeouts longer (and more configurable). This should sort that out.

--- 
Note that there is a bit of tidying up of the codebase along the way. Some of these refactoring/fix commits should be cherry-picked to the patch branch...

### CODE REVIEW

File: packages/compiler-cli/ngcc/src/packages/configuration.ts [hunk 1]
```diff
@@ -20,7 +20,27 @@ export interface NgccProjectConfig<T = NgccPackageConfig> {
   /**
    * The packages that are configured by this project config.
    */
-  packages: {[packagePath: string]: T};
+  packages?: {[packagePath: string]: T};
+  /**
+   * Options that control how the locking the process is handled.
```
Comment (JoostK) [inline packages/compiler-cli/ngcc/src/packages/configuration.ts:25]:
```suggestion
   * Options that control how locking the process is handled.
```

File: packages/compiler-cli/ngcc/src/packages/configuration.ts [hunk 2]
```diff
@@ -20,7 +20,27 @@ export interface NgccProjectConfig<T = NgccPackageConfig> {
...
+   */
+  locking?: ProcessLockingConfiguration;
+}
+
+/**
+ * Options that control how the locking the process is handled.
```
Comment (JoostK) [inline packages/compiler-cli/ngcc/src/packages/configuration.ts:31]:
```suggestion
 * Options that control how locking the process is handled.
```

File: packages/compiler-cli/ngcc/src/packages/configuration.ts [hunk 3]
```diff
@@ -126,12 +146,18 @@ export const DEFAULT_NGCC_CONFIG: NgccProjectConfig = {
       },
     },
   },
+  locking: {
+    retryDelay: 500,
```
Comment (JoostK) [inline packages/compiler-cli/ngcc/src/packages/configuration.ts:150]: This is 500ms, not 50ms.

**Review by JoostK — CHANGES_REQUESTED** [review]
There is something off in the ms calculations in the latest commit. 50x50ms is not 25s, but just 2.5s if my math is not mistaken. It may currently work out well enough because the interval is actually set to 500ms, not 50ms.

File: packages/compiler-cli/ngcc/src/packages/configuration.ts (hunk 3 above)
Comment (petebacondarwin) [reply packages/compiler-cli/ngcc/src/packages/configuration.ts:150]: Yes in the earlier commits I mistakenly swapped  the `retryAttempts` with the `retryDelay` values. Should be resolved in the fixup commits.

**Review by JoostK — APPROVED** [review]
LGTM, except for the commit message of "refactor(ngcc): change async locker timeout to 250 secs" as its numbers are incorrect.

**Comment by petebacondarwin** [comment]
Thanks @JoostK - reworded the commit message.

File: packages/compiler-cli/ngcc/main-ngcc.ts [hunk 4]
```diff
@@ -24,7 +24,7 @@ if (require.main === module) {
       process.exitCode = 0;
     } catch (e) {
       console.error(e.stack || e.message);
-      process.exit(1);
+      process.exit(e.code || 1);
```
Comment (gkalpak) [inline packages/compiler-cli/ngcc/main-ngcc.ts:27]: [process.exit()][1] must be called with an integer, but `e.code` can be string for [errors thrown in Node.js][2].

[1]: https://nodejs.org/api/process.html#process_process_exit_code
[2]: https://nodejs.org/api/errors.html#errors_error_code

File: packages/compiler-cli/ngcc/src/packages/configuration.ts [hunk 5]
```diff
@@ -170,21 +196,33 @@ export class NgccConfiguration {
...
   }
 
+  /**
+   * Get the configuration options for locking the ngcc process.
+   */
+  getLockingConfig(): Required<ProcessLockingConfiguration> {
+    const retryAttempts =
+        this.projectConfig.locking.retryAttempts || this.defaultConfig.locking.retryAttempts!;
+    const retryDelay =
+        this.projectConfig.locking.retryDelay || this.defaultConfig.locking.retryDelay!;
```
Comment (gkalpak) [inline packages/compiler-cli/ngcc/src/packages/configuration.ts:206]: What is someone wants to set `retryAttempts` or `retryDelay` to 0? :grin:

**Review by gkalpak — CHANGES_REQUESTED** [review]
Nice 😄

File: packages/compiler-cli/ngcc/main-ngcc.ts (hunk 4 above)
Comment (petebacondarwin) [reply packages/compiler-cli/ngcc/main-ngcc.ts:27]: Good point. I'll add an "is numeric" check.

File: packages/compiler-cli/ngcc/src/packages/configuration.ts (hunk 5 above)
Comment (petebacondarwin) [reply packages/compiler-cli/ngcc/src/packages/configuration.ts:206]: OK OK

**Review by gkalpak — APPROVED** [review]

**Review by IgorMinar — APPROVED** [review]
LGTM! Thanks!

Reviewed-for: integration-tests

**Comment by AndrewKushnir** [comment]
Hi @petebacondarwin, could you please rebase this PR and resolve conflicts? Thank you.

**Pull request closed without merging**